### PR TITLE
[POC] Filter raw values

### PR DESCRIPTION
**Pull request by ahocquard**
**Description (for Contributor and Core Developer)**
Goal is to filter values that does not exist. For now only attributes and attribute option codes.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Component/Catalog/ValuesFilter/ExistingAttributeFilter.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * @param array $attributeCodes
+     *
+     * @return array indexed by attribute code
+     *
+     */
+    private function getExistingAttributeCodes(array $attributeCodes): array
+    {
+        $sql = <<<SQL
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/ValuesFilter/ExistingAttributeFilter.php]: Sql request in the component?

Comment (ahocquard) [reply src/Pim/Component/Catalog/ValuesFilter/ExistingAttributeFilter.php]: You're right, this class should be in the infra, but as it's a POC I kept it all in one directory

**Comment by jjanvier** [comment]
I agree with Arnaud and what you said yesterday. This doesn't seem to be good.
As AL suggested, why not a `RawValueCollection`. You can try... But I think it will be difficult to create to this object completely correctly in one shot (as you'll need to call the database). So maybe you'll need also a factory :( Which means you'll end up with another layer doing the same thing that we have today. 

So maybe the first right baby step is: instead of doing the database checks in all the unitary factories, try to do it at one place, in the `ValueCollectionFactory` (or in a delegated object). This will not change or enhance the design, but could improve a bit performances. Once done, you could try to work on another construction work or POC: try to remove the hard links with Attribute, Locale and Channel and see again if it improves performances. Or maybe you could try at this step to create the `RawValues` object and see if it improves the design.

Viewing your PR made me think about 2 things:
- I don't think we should do big bang refactorings on this topic. This will lead to nowhere.
- I don't think we should try to improve the performances and the design at the same time. We should do it step by step, one after the other.

Maybe the best would be to know where we want to go, the ideal goal (which is super hard, and require brainstormings). And try to find the path towards this ideal.

File: src/Pim/Component/Catalog/ValuesFilter/ExistingSimpleSelectOptionFilter.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * @param array $rawValuesList
+     * @param array $simpleSelectAttributeCodes
+     *
+     * @return array
+     */
+    private function getOptionCodes(array $rawValuesList, array $simpleSelectAttributeCodes)
+    {
+        $optionCodes = [];
+        foreach ($rawValuesList as $rawValues) {
```
[the remark this answers is not in the document]
Comment (anaelChardan) [reply src/Pim/Component/Catalog/ValuesFilter/ExistingSimpleSelectOptionFilter.php]: Totally agree, I already propose it :D, I would be very happy to use it ;)

File: src/Pim/Component/Catalog/ValuesFilter/ExistingSimpleSelectOptionFilter.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                    AttributeTypes::OPTION_SIMPLE_SELECT
+                ]
+            ],
+            [
+                'attribute_codes' => \Doctrine\DBAL\Connection::PARAM_STR_ARRAY
+            ]
+        )->fetchAll();
+
+        $existingAttributeCodes = [];
+        foreach ($rows as $row) {
```
[the remark this answers is not in the document]
Comment (anaelChardan) [reply src/Pim/Component/Catalog/ValuesFilter/ExistingSimpleSelectOptionFilter.php]: or a map ;)

**Comment by ahocquard** [comment]
@jjanvier I already did (more or less) this thing to improve hydration with kinf of `cache warmup`:
https://github.com/akeneo/pim-community-dev/pull/7588

**Comment by ahocquard** [comment]
I close it as it's not very clean.
Also, I did a quick benchmark of PHP and I can create 2,5 millions of objects in 1 second on my laptopt (core i5). 

It means that creating the ValueCollection costs (almost) nothing.
But filtering by doing several hundreds SQL requests costs a lot ( and it was the goal of my PR to batch the queries).

We have to fix it by (still) batching requests and filtering directly the value collection.
The added value is that we can index things correctly in the value collection to filter it quickly.

**Pull request closed without merging**